axlearn/common/learner.py

```diff
@@ -444,6 +444,153 @@ def _mask_tree(tree: dict, *, keep: dict) -> dict:
     )
 
 
+class MetricsAccumulationOp(NamedTuple):
+    microbatches: int
+
+    def aggregrate(self, x, buffer):
+        raise NotImplementedError(self)
+
+    def normalize(self, buffer):
+        raise NotImplementedError(self)
+
+
+class ArithmeticMeanStrategy(MetricsAccumulationOp):
+    def aggregrate(self, x, buffer):
+        return buffer + x
+
+    def normalize(self, buffer):
+        return buffer / self.microbatches
+
+
+class GeometricMeanStrategy(MetricsAccumulationOp):
+    def aggregrate(self, x, buffer):
+        return buffer * x
+
+    def normalize(self, buffer):
+        return buffer ** (-self.microbatches)
+
+
+class AddStrategy(MetricsAccumulationOp):
+    def aggregrate(self, x, buffer):
+        return buffer + x
+
+    def normalize(self, buffer):
+        return buffer
+
+
+class AccumulatedLearner(Learner):
+    """Learner module that uses gradient accumulation"""
+
+    @config_class
+    class Config(Learner.Config):
+        """Configures Learner."""
+
+        # The number of microbatches to accumulate per optimizer step.
+        microbatches: Required[int] = REQUIRED
+        # tuple of key-value pairs specifying custom aggregation and normalization
+        # for a specific metric
+        metrics_accumulation_key_ops: Sequence[Dict[str, Optional[MetricsAccumulationOp]]] = []
+        gradient_dtype: Optional[jnp.dtype] = jnp.bfloat16
+
+    def forward_and_backward(
+        self, *, fn: ForwardFn, inputs: NestedTensor, opt_params: NestedOptParam
+    ) -> ForwardBackwardOutputs:
+        """Run forward and backward for multiple microbatches and accumulate gradients
+        and metrics into buffers.
+
+        Args:
+            fn (ForwardFn): Model forward pass
+            inputs (NestedTensor): input microbatches with a leading microbatch axis
+            opt_params (NestedOptParam): params which don't need gradients
+
+        Returns:
+            ForwardBackwardOutputs: pytree containing gradients and metrics
+        """
```

Comment:
I wonder if instead of having a separate learner for microbatching, it would be more flexible to have a generic way of wrapping a ForwardFn so that it uses Jax.lax.map to run the microbatches. Beyond avoiding the need to add a new learner, it would also allow for other microbetching uses outside of learner, eg inference or in second order optimizers.

Reply:
jax.lax.map gives no guarantees of sequential execution of microbatches which is the key quality of gradient accumulation.